src/town_cmd.cpp

```diff
@@ -1199,6 +1199,72 @@ static bool GrowTownWithBridge(const Town *t, const TileIndex tile, const DiagDi
 	return false;
 }
 
+/**
+ * Grows the town with a tunnel.
+ *  First we check if a tunnel is reasonable.
+ *  If so we check if we are able to build it.
+ *
+ * @param t The current town
+ * @param tile The current tile
+ * @param tunnel_dir The valid direction in which to grow a tunnel
+ * @return true if a tunnel has been built, else false
+ */
+static bool GrowTownWithTunnel(const Town* t, const TileIndex tile, const DiagDirection tunnel_dir)
+{
+	assert(tunnel_dir < DIAGDIR_END);
+
+	Slope slope = GetTileSlope(tile);
+
+	/* Only consider building a tunnel if the starting tile is sloped properly. */
+	if (slope != InclinedSlope(tunnel_dir)) return false;
+
+	const int delta = TileOffsByDiagDir(tunnel_dir);
+	int max_tunnel_length = 0;
+
+	/* There are two conditions for building tunnels: Under a mountain and under an obstruction. */
+	if (CanRoadContinueIntoNextTile(t, tile, tunnel_dir)) {
+		/* Only tunnel under a mountain if the slope continues for at least 4 tiles, signifying a large mountain. */
+		TileIndex slope_tile = tile;
+		for (uint8 tiles = 0; tiles < 4; tiles++) {
+			slope = GetTileSlope(slope_tile);
+			if (slope != InclinedSlope(tunnel_dir) && !IsSteepSlope(slope) && !IsSlopeWithOneCornerRaised(slope)) return false;
+			slope_tile += delta;
+		}
```

Comment:
It feels like we are doing the same twice. Here we check the first 4 tiles, and later we check all of them. All with slightly different logic (here you check slopes, but when you check if the tunnel can be build, you only check if you get back to the same height).

How about instead of making this more complicated than needed, why not introduce a `min_tunnel_length` ? That would mean you only have to set it to 4 here, and check after the `do` loop that it is at least `min_tunnel_length` before continueing?

Reply:
The slope check is separate because it has nothing to do with tunnel length, rather it is checking for a continuous slope. If the slope is not continuous and the road could level out and build a perpendicular road, the tunnel is rejected. I want to prioritize town growth and tunnel only if the hill can't be developed.

Reply:
Given it's clearly unclear, perhaps a small comment?